File: content_tool/agents/outline.py
from datetime import date
from typing import Literal
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncSession

from content_tool import prompts_store
from content_tool.db.models import FetchedArticle, GapAnalysisRow, OutlineRow, Run
from content_tool.gemini.client import GeminiClient
from content_tool.gemini.prompt_context import PromptMeta, set_prompt_meta
from content_tool.models.outline import Outline
from content_tool.policy.personas import load_persona


async def build_system_prompt(
    today: date,
    start_mode: Literal["refresh", "create"] = "refresh",
    *,
    voice_slug: str = "bowtie-editor",
    session: AsyncSession,
) -> str:
    """Render the outline system prompt for ``voice_slug`` (the run's voice).

    ``start_mode == "create"``: slot the create-mode body into the
    ``{create_mode_block}`` placeholder so the LLM gets the create-mode brief
    *before* the refresh-mode instructions in the template.

    ``start_mode == "refresh"`` (default): the placeholder is replaced with
    an empty string and only the refresh-mode body applies — identical to
    the pre-Task-4 behaviour.

    Both the ``outline_rewrite_mode`` and ``outline_create_mode`` templates
    resolve under ``voice_slug`` (falling back to __shared__ / bundled file).
    """
    block = (
        (
            await prompts_store.get_assembled(
                "outline_create_mode", voice_slug=voice_slug, session=session
            )
        ).rstrip()
        if start_mode == "create"
        else ""
    )
    template = await prompts_store.get_assembled(
        "outline_rewrite_mode", voice_slug=voice_slug, session=session
    )
    row = await prompts_store.get_template_row(
        "outline_rewrite_mode", voice_slug=voice_slug, session=session
    )
    if row is not None:
        set_prompt_meta(PromptMeta(
            template_id=row.template_id, voice_slug=row.voice_slug, sha256=row.sha256
        ))
    # Locale/brand tokens (mirror writer.build_system_prompt). The create-mode
    # block is injected FIRST so any {output_language}/{brand_name}/{market}
    # tokens it carries are interpolated by the replaces below. HK-ZH defaults
    # equal the old literals → byte-identical for bowtie-editor.
    loc = (await load_persona(voice_slug, session=session)).locale
    return (
        template.replace("{today_date}", today.isoformat())
        .replace("{create_mode_block}", block)
        .replace("{brand_name}", loc.brand_name)
        .replace("{output_language}", loc.output_language)
        .replace("{market}", loc.market)
    )


def build_user_prompt(
    *,
    gap_analysis_payload: dict,
    existing_markdown: str,
    chosen_route: str,
    acf_adv_id: int,
    acf_widget_id: int,
) -> str:
    import json as _j

    return (
        f"chosen_route: {chosen_route}\n"
        f"acf_adv_id: {acf_adv_id}\n"
        f"acf_widget_id: {acf_widget_id}\n\n"
        f"# gap_analysis\n{_j.dumps(gap_analysis_payload, ensure_ascii=False)}\n\n"
        f"# existing_article_markdown\n{existing_markdown}"
    )


def build_user_prompt_create_mode(
    *,
    topic: str,
    keywords: list[str],
    target_audience: str | None,
    acf_adv_id: int,
    acf_widget_id: int,
    edit_note: str | None = None,
) -> str:
    """Create-mode user prompt — there is no fetched article and no gap
    analysis to feed the model, just the brief from the operator (Front III)
    or the promoted topic candidate (Front II).
    """
    kw = ", ".join(keywords) if keywords else "(無)"
    audience = target_audience or "(未指定)"
    note_block = (
        f"編輯指示（最優先）：{edit_note}\n"  # noqa: RUF001
        if edit_note
        else ""
    )
    # Full-width colons are intentional — the prompt is rendered for CJK
    # readers and the model is conditioned on this style throughout the
    # prompts/ folder.
    return (
        f"主題：{topic}\n"  # noqa: RUF001
        f"關鍵字：{kw}\n"  # noqa: RUF001
        f"目標讀者：{audience}\n"  # noqa: RUF001
        f"acf_adv_id: {acf_adv_id}\n"
        f"acf_widget_id: {acf_widget_id}\n"
        f"{note_block}"
    )


async def run_outline(
    *,
    session: AsyncSession,
    gemini: GeminiClient,
    run_id: UUID,
    today: date,
) -> Outline:
    run = (await session.execute(select(Run).where(Run.run_id == run_id))).scalar_one()
    start_mode: Literal["refresh", "create"] = (
        "create" if run.start_mode == "create" else "refresh"
    )

    sys_prompt = await build_system_prompt(
        today, start_mode, voice_slug=run.persona, session=session
    )

    if start_mode == "create":
        user_prompt = build_user_prompt_create_mode(
            topic=run.topic,
            keywords=list(run.keywords or []),
            target_audience=run.target_audience,
            acf_adv_id=run.acf_adv_id,
            acf_widget_id=run.acf_widget_id,
            edit_note=run.edit_note,
        )
    else:
        fa = (
            await session.execute(select(FetchedArticle).where(FetchedArticle.run_id == run_id))
        ).scalar_one()
        ga = (
            await session.execute(select(GapAnalysisRow).where(GapAnalysisRow.run_id == run_id))
        ).scalar_one()
        user_prompt = build_user_prompt(
            gap_analysis_payload=ga.payload,
            existing_markdown=fa.markdown,
            chosen_route=run.chosen_route or "small_refresh",
            acf_adv_id=run.acf_adv_id,
            acf_widget_id=run.acf_widget_id,
        )

    result = await gemini.generate(
        agent="outline",
        system_prompt=sys_prompt,
        user_prompt=user_prompt,
        response_schema=Outline.model_json_schema(),
        tools=[],
    )
    outline = Outline.model_validate(result.parsed)

    # Upsert so re-running the node (e.g. after a restart) refreshes the outline
    # instead of violating outlines_pkey. Preserve any prior human edits.
    payload = outline.model_dump()
    stmt = (
        pg_insert(OutlineRow)
        .values(run_id=run_id, payload=payload, edited_by_human=False)
        .on_conflict_do_update(
            index_elements=[OutlineRow.run_id],
            set_={"payload": payload},
        )
    )
    await session.execute(stmt)
    await session.commit()
    return outline

File: deploy/cloudflare-workers/src/agents/outline.ts
/**
 * Outline agent — TypeScript port of `content_tool/agents/outline.py`.
 *
 * Supports both "create" and "refresh" start modes. The create-mode path is
 * the primary focus; the refresh-mode path is also ported for completeness.
 */

import type { Sql } from "postgres";
import { toJsonb } from "../db/serialize";
import { getAssembled, substitute } from "../prompts/store";
import { loadPersona } from "./persona";
import type { GeminiClient, ThoughtCallback } from "../gemini/types";
import { OUTLINE_SCHEMA, type Outline } from "./schemas";

// ---------------------------------------------------------------------------
// Input
// ---------------------------------------------------------------------------

export interface OutlineInput {
  runId: string;
  startMode: "create" | "refresh";
  /** The run's voice (persona slug); resolves the prompt template under it. */
  voiceSlug: string;
  topic: string;
  keywords: string[];
  targetAudience?: string | null;
  acfAdvId: number;
  acfWidgetId: number;
  editNote?: string | null;
  todayDate: string; // YYYY-MM-DD
  /** Refresh mode: gap-analysis JSON payload */
  gapAnalysisPayload?: object | null;
  /** Refresh mode: fetched article markdown */
  existingMarkdown?: string | null;
  /** Refresh mode: chosen route string */
  chosenRoute?: string | null;
  onThought?: ThoughtCallback;
}

// ---------------------------------------------------------------------------
// Token usage result
// ---------------------------------------------------------------------------

export interface OutlineTokens {
  tokensIn: number;
  tokensOut: number;
  thinkingTokens: number;
  latencyMs: number;
}

// ---------------------------------------------------------------------------
// System prompt assembly — mirrors Python `build_system_prompt`
//
// For "create": fetch outline_create_mode, rstrip, slot into {create_mode_block}.
// For "refresh": replace {create_mode_block} with "".
// Then substitute {today_date} with todayDate.
// ---------------------------------------------------------------------------

async function buildSystemPrompt(
  sql: Sql,
  startMode: "create" | "refresh",
  todayDate: string,
  voiceSlug: string,
): Promise<string> {
  const block =
    startMode === "create"
      ? (await getAssembled(sql, "outline_create_mode", voiceSlug)).replace(/\n+$/, "")
      : "";

  const template = await getAssembled(sql, "outline_rewrite_mode", voiceSlug);

  // Inject the create-mode block FIRST, then interpolate locale/brand tokens
  // (mirror outline.py / writer.ts). String.replace is first-match-only in JS
  // but these tokens recur, so use replaceAll. HK-ZH defaults equal the old
  // literals → byte-identical for bowtie-editor.
  const { locale: loc } = await loadPersona(sql, voiceSlug);
  return substitute(template, {
    today_date: todayDate,
    create_mode_block: block,
  })
    .replaceAll("{brand_name}", loc.brandName)
    .replaceAll("{output_language}", loc.outputLanguage)
    .replaceAll("{market}", loc.market);
}

// ---------------------------------------------------------------------------
// User prompt — create mode
//
// Mirrors Python `build_user_prompt_create_mode` exactly.
// Full-width colons are intentional — the prompts use CJK style throughout.
// ---------------------------------------------------------------------------

export function buildUserPromptCreateMode(opts: {
  topic: string;
  keywords: string[];
  targetAudience: string | null | undefined;
  acfAdvId: number;
  acfWidgetId: number;
  editNote?: string | null;
}): string {
  const kw = opts.keywords.length > 0 ? opts.keywords.join(", ") : "(無)";
  const audience = opts.targetAudience ?? "(未指定)";
  const noteBlock = opts.editNote ? `編輯指示（最優先）：${opts.editNote}\n` : "";

  return (
    `主題：${opts.topic}\n` +
    `關鍵字：${kw}\n` +
    `目標讀者：${audience}\n` +
    `acf_adv_id: ${opts.acfAdvId}\n` +
    `acf_widget_id: ${opts.acfWidgetId}\n` +
    noteBlock
  );
}

// ---------------------------------------------------------------------------
// User prompt — refresh mode
//
// Mirrors Python `build_user_prompt`.
// ---------------------------------------------------------------------------

export function buildUserPromptRefresh(opts: {
  gapAnalysisPayload: object;
  existingMarkdown: string;
  chosenRoute: string;
  acfAdvId: number;
  acfWidgetId: number;
}): string {
  return (
    `chosen_route: ${opts.chosenRoute}\n` +
    `acf_adv_id: ${opts.acfAdvId}\n` +
    `acf_widget_id: ${opts.acfWidgetId}\n\n` +
    `# gap_analysis\n${JSON.stringify(opts.gapAnalysisPayload)}\n\n` +
    `# existing_article_markdown\n${opts.existingMarkdown}`
  );
}

// ---------------------------------------------------------------------------
// Main export
// ---------------------------------------------------------------------------

/**
 * Run the outline agent for a single run.
 *
 * Assembles the system + user prompt, calls Gemini with structured output,
 * and upserts the result into `content_tool.outlines` — preserving any prior
 * human edits (edited_by_human / human_edits columns are never overwritten).
 */
export async function runOutline(
  sql: Sql,
  gemini: GeminiClient,
  input: OutlineInput,
): Promise<{ outline: Outline; tokens: OutlineTokens }> {
  const systemPrompt = await buildSystemPrompt(
    sql,
    input.startMode,
    input.todayDate,
    input.voiceSlug,
  );

  let userPrompt: string;
  if (input.startMode === "create") {
    userPrompt = buildUserPromptCreateMode({
      topic: input.topic,
      keywords: input.keywords,
      targetAudience: input.targetAudience,
      acfAdvId: input.acfAdvId,
      acfWidgetId: input.acfWidgetId,
      editNote: input.editNote,
    });
  } else {
    userPrompt = buildUserPromptRefresh({
      gapAnalysisPayload: input.gapAnalysisPayload ?? {},
      existingMarkdown: input.existingMarkdown ?? "",
      chosenRoute: input.chosenRoute ?? "small_refresh",
      acfAdvId: input.acfAdvId,
      acfWidgetId: input.acfWidgetId,
    });
  }

  const result = await gemini.generate({
    agent: "outline",
    systemPrompt,
    userPrompt,
    responseSchema: OUTLINE_SCHEMA as Record<string, unknown>,
    tools: [],
    onThought: input.onThought,
  });

  // Cast the parsed response — Gemini returns a plain object matching Outline's
  // shape when responseSchema is provided. Route through `unknown` to satisfy
  // the compiler since `GeminiResult.parsed` is `Record<string, unknown>`.
  const outline = result.parsed as unknown as Outline;

  // UPSERT into content_tool.outlines.
  // On conflict (run_id already exists — e.g. node restart), update only the
  // payload column. edited_by_human and human_edits are left untouched so that
  // any prior human review is not clobbered.
  //
  // `sql.json(...)` sends a NATIVE jsonb param (type OID 3802, serialized once)
  // so the column stores a jsonb OBJECT — not a double-encoded string scalar.
  await sql`
    INSERT INTO content_tool.outlines (run_id, payload, edited_by_human)
    VALUES (${input.runId}, ${toJsonb(sql, outline)}, ${false})
    ON CONFLICT (run_id) DO UPDATE
      SET payload = excluded.payload
  `;

  return {
    outline,
    tokens: {
      tokensIn: result.tokensIn,
      tokensOut: result.tokensOut,
      thinkingTokens: result.thinkingTokens,
      latencyMs: result.latencyMs,
    },
  };
}
